cf/data/data.py

```diff
@@ -8811,32 +8812,27 @@ def equals(
         other_Units = other.Units
         if self_Units != other_Units:
             logger.info(
-                "{}: Different Units ({!r}, {!r}".format(
+                "{}: Different Units ({!r}, {!r})".format(
                     self.__class__.__name__, self.Units, other.Units
                 )
             )
             return False
 
-        config = self.partition_configuration(readonly=True)
+        # other.to_memory()  # TODODASK is this still required?
 
-        other.to_memory()
+        self_dx = self._get_dask()
+        other_dx = other._get_dask()
 
-        for partition in self.partitions.matrix.flat:
-            partition.open(config)
-            array0 = partition.array
-            array1 = other[partition.indices].varray
-            partition.close()
-
-            if not _numpy_allclose(
-                array0, array1, rtol=float(rtol), atol=float(atol)
-            ):
-                logger.info(
-                    "{0}: Different array values (atol={1}, "
-                    "rtol={2})".format(self.__class__.__name__, atol, rtol)
-                )
+        # Finally check that corresponding elements are equal to a tolerance
+        if not da.allclose(
```

Comment:
Not sure if `da.allcose` is enough here, e.g. the following gives `True` in dask (and numpy), but we want it to be `False`:
```python
>>> from dask import array as da
>>> import numpy as np
>>> dx = da.from_array(np.ma.array([1, 999, 3], mask=[0, 1, 0]))
>>> dy = da.from_array(np.ma.array([999, 2, 3], mask=[1, 0, 0]))
>>> da.allclose(dx, dy).compute()
True
```
This is why we use the (terribly named) `cf.functions._numpy_allclose` function instead, which also accounts for non-numeric data types.

Daskifying  `cf.functions._numpy_allclose` will need a little thought to see if we can avoid computing the two operands multiple times each (once to get the mask, and once to compare the data).

Reply:
Ah, OK, good points. It sounds like, firstly, my unit test needs to be more thorough so it tests for cases like that, so that currently with the change set here it will fail, then I need to fix the `equals` in the way you state is required.

> Daskifying cf.functions._numpy_allclose will need a little thought to see if we can avoid computing the two operands multiple times each (once to get the mask, and once to compare the data).

I'll take a further look and have a think about a good approach, though I think we discussed a similar challenge back in March or around then, so that might guide us as to a good solution. I'll look through my notes to see what we thought back then and report back briefly and/or add a new commit with a suggestion.

Reply:
> This is why we use the (terribly named) cf.functions._numpy_allclose function instead

I realised now (whilst having a more detailed look) that I assumed `_numpy_allclose` was an alias to `numpy.allclose` and hence the direct and seemingly obvious conversion I suggested to `da.allclose`. (So yes, it is somewhat more complicated than I thought to migrate this.) I think, even though those `_numpy_*` aliases have been removed for the `lama-to-dask` branch, it would be best to rename `_numpy_allclose` since the name is rather confusing.

I'll get my migration fix suggest committed and in the meantime if you have any thoughts about a less confusing name, do let me know. It does seem difficult to name, though. Maybe `_inclusive_allclose` or something similar? (It isn't too important since it is an internal function.)